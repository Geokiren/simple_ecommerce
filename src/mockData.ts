export const products = [
  {
    id: 1,
    title: 'Product 1',
    description: 'Product 1 description',
    category: 'shoe',
    price: 10.99,
    discountPercentage: 10,
    rating: 4.5,
    stock: 0,
    tags: [],
    brand: '',
    sku: '',
    weight: 0,
    dimensions: {
      width: 0,
      height: 0,
      depth: 0
    },
    warrantyInformation: '',
    shippingInformation: '',
    availabilityStatus: '',
    reviews: [
      {
        reviewerName: 'John Doe',
        reviewerEmail: 'vqzJZ@example.com',
        rating: 5,
        comment: 'Great product!',
        date: '2022-01-01'
      }
    ],
    returnPolicy: '',
    minimumOrderQuantity: 0,
    meta: {
      createdAt: '',
      updatedAt: '',
      barcode: '',
      qrCode: ''
    },
    thumbnail: '',
    images: [],
    quantity: 1
  },
  {
    id: 2,
    title: 'Product 2',
    description: '',
    category: '',
    price: 20.49,
    discountPercentage: 0,
    rating: 0,
    stock: 0,
    tags: [],
    brand: '',
    sku: '',
    weight: 0,
    dimensions: {
      width: 0,
      height: 0,
      depth: 0
    },
    warrantyInformation: '',
    shippingInformation: '',
    availabilityStatus: '',
    reviews: [],
    returnPolicy: '',
    minimumOrderQuantity: 0,
    meta: {
      createdAt: '',
      updatedAt: '',
      barcode: '',
      qrCode: ''
    },
    thumbnail: '',
    images: [],
    quantity: 1
  }
];
