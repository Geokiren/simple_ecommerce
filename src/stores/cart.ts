import { ref, computed } from 'vue';
import { defineStore } from 'pinia';

import type { Product } from '../types';

export const useCartStore = defineStore('cart', () => {
  const cartProducts = ref<Product[]>([]);

  const cartCounter = computed(() =>
    cartProducts.value.reduce((acc, curr) => acc + (curr.quantity ?? 0), 0)
  );

  const totalPrice = computed(() =>
    parseFloat(
      cartProducts.value
        .reduce((acc, curr) => acc + curr.price * (curr.quantity ?? 0), 0)
        .toFixed(2)
    )
  );

  const addToBasket = (product: Product) => {
    const foundProduct = cartProducts.value.find((p) => p.id === product.id);

    if (!foundProduct) {
      const newProduct = { ...product, quantity: 1 };
      cartProducts.value.push(newProduct);
    } else {
      foundProduct.quantity = (foundProduct.quantity ?? 0) + 1;
    }

    cartProducts.value = [...cartProducts.value];
  };

  const removeFromBasket = (product: Product) => {
    cartProducts.value = cartProducts.value.filter((p) => p.id !== product.id);
  };

  return { cartProducts, cartCounter, totalPrice, addToBasket, removeFromBasket };
});
