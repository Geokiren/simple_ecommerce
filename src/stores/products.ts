import { ref, computed, watch } from 'vue';
import { defineStore } from 'pinia';

import {
  getCategories,
  searchProducts,
  getProducts,
  getProductsByCategory,
  getProduct
} from '@/services/products';

import type { Product } from '../types';
import type { Category } from '../types/category';

interface IPaging {
  limit: number;
  skip: number;
}

interface ISortBy {
  sortBy: string;
  order: string;
}

export const useProductsStore = defineStore('products', () => {
  const categories = ref<Category[]>([]);
  const products = ref<Product[]>([]);

  const paging = ref<IPaging>({
    limit: 30,
    skip: 0
  });

  const total = ref(0);

  const searchTerm = ref<string>('');

  const sorting = ref<ISortBy>({
    sortBy: 'title',
    order: 'asc'
  });

  const activeCategory = ref<Category | null>(null);

  const loading = ref(false);

  const filters = computed(() => {
    const sortBy = sorting.value.sortBy ? `&sortBy=${sorting.value.sortBy}` : '';
    const order = sortBy ? `&order=${sorting.value.order}` : '';
    const pagingParams = `&limit=${paging.value.limit}&skip=${paging.value.skip}`;

    return `?${new URLSearchParams(sortBy + order + pagingParams).toString()}`;
  });

  const setSearchTerm = (term: string) => {
    searchTerm.value = term;
  };

  const setActiveCategory = (category: Category | null) => {
    activeCategory.value = category;
  };

  const setSorting = (value: string, property: 'sortBy' | 'order') => {
    sorting.value[property] = value;
  };

  const hasMorePages = computed(() => {
    return products.value.length < total.value;
  });

  const incrementPage = () => {
    if (hasMorePages.value) {
      paging.value = { ...paging.value, skip: paging.value.skip + paging.value.limit };
    }
  };

  function debounce<T extends (...args: any[]) => any>(fn: T, delay: number) {
    let timeoutId: ReturnType<typeof setTimeout> | null = null;

    return function (this: any, ...args: Parameters<T>) {
      if (timeoutId !== null) {
        clearTimeout(timeoutId);
      }

      timeoutId = setTimeout(() => {
        fn.apply(this, args);
        timeoutId = null;
      }, delay);
    };
  }

  const debouncedSearch = debounce(async () => {
    await fetchSearchResults();
  }, 300);

  watch(searchTerm, async (value: string) => {
    if (!value) {
      products.value = [];
      fetchProducts();
      return;
    }

    if (value.length >= 3) {
      products.value = [];
      debouncedSearch();
    }
  });

  watch(activeCategory, (value: Category | null) => {
    if (!value) {
      products.value = [];
      fetchProducts();
      return;
    }

    products.value = [];
    fetchProductsByCategory();
  });

  watch(paging, (newValue, oldValue) => {
    if (activeCategory.value) {
      if (newValue.limit !== oldValue.limit) products.value = [];
      fetchProductsByCategory();
      return;
    }

    if (newValue.limit !== oldValue.limit) products.value = [];
    fetchProducts();
  });

  watch(
    sorting,
    () => {
      if (activeCategory.value) {
        products.value = [];
        fetchProductsByCategory();
      } else {
        products.value = [];
        fetchProducts();
      }
    },
    { deep: true }
  );

  const fetchCategories = async (): Promise<Category[]> => {
    try {
      const categories: Category[] = await getCategories();
      return categories;
    } catch (error) {
      console.error(error);
      throw error;
    }
  };

  const fetchSearchResults = async (): Promise<{ products: Product[]; total: number }> => {
    loading.value = true;

    try {
      const data = await searchProducts(searchTerm.value);

      products.value = data.products;
      total.value = data.total;

      return { products: data.products, total: data.total };
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      loading.value = false;
    }
  };

  const fetchProducts = async (): Promise<{ products: Product[]; total: number }> => {
    loading.value = true;

    try {
      const data = await getProducts(filters.value);

      products.value = [...products.value, ...data.products];
      total.value = data.total;

      return { products: data.products, total: data.total };
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      loading.value = false;
    }
  };

  const fetchProductsByCategory = async (returnData?: boolean, category?: string) => {
    if (!category && !activeCategory.value) {
      console.error('No category provided');
      return;
    }

    loading.value = true;

    try {
      const categoryToUse = category || activeCategory.value?.slug || '';

      const data = await getProductsByCategory(categoryToUse, filters.value);

      if (returnData) {
        return data?.products || [];
      } else {
        products.value = [...products.value, ...data.products];
        total.value = data.total;
      }
    } catch (error) {
      console.error(error);
    } finally {
      loading.value = false;
    }
  };

  const fetchProduct = async (id: string) => {
    try {
      const product = await getProduct(id);
      return product;
    } catch (error) {
      console.error(error);
    }
  };

  return {
    categories,
    activeCategory,
    products,
    searchTerm,
    loading,
    fetchCategories,
    fetchSearchResults,
    fetchProductsByCategory,
    fetchProducts,
    fetchProduct,
    setSearchTerm,
    setActiveCategory,
    setSorting,
    incrementPage
  };
});
