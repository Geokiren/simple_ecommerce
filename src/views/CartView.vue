<template>
  <section class="max-w-screen-2xl mx-auto px-6 py-6 flex flex-col gap-4" data-test="cart-view">
    <h1 class="text-2xl text-center  font-bold px-4 pt-4 pb-2 w-full border-b">Your Bag</h1>

    <ProductsCart v-if="totalPrice > 0" />

    <div v-else class="text-sm text-center">Your bag is empty</div>

    <div class="p-4 flex items-center justify-between">
      <div class="font-bold">Subtotal:</div>

      <div data-test="total-price" >{{ totalPrice }}â‚¬</div>
    </div>
  </section>
</template>

<script setup lang="ts">
  import { useCartStore } from '@/stores/cart';
  import { storeToRefs } from 'pinia';

  import ProductsCart from '@/components/products/ProductsCart.vue';

  const cartStore = useCartStore();
  const { totalPrice } = storeToRefs(cartStore);
</script>