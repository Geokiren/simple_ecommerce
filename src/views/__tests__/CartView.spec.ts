import { mount } from '@vue/test-utils';
import { describe, expect, it, beforeEach } from 'vitest';
import { useCartStore } from '@/stores/cart';
import { createPinia } from 'pinia';

import CartView from '../CartView.vue';

import { products } from '../../mockData';

describe('CartView', () => {
  const cartStore = useCartStore(createPinia());

  beforeEach(() => {
    cartStore.cartProducts = [];
  });

  it('renders the cart view with a non-empty bag', () => {
    cartStore.cartProducts.push(products[0]);

    const wrapper = mount(CartView);

    expect(wrapper.find('[data-test="cart-view"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="total-price"]').text()).toBe('10.99€');
  });

  it('renders the cart view with an empty bag', () => {
    const wrapper = mount(CartView);

    expect(wrapper.find('[data-test="cart-view"]').exists()).toBe(true);
    expect(wrapper.find('div.text-sm.text-center').text()).toBe('Your bag is empty');
  });

  it('renders the subtotal and total price correctly', () => {
    cartStore.cartProducts.push(products[0], products[1]);

    const wrapper = mount(CartView);

    expect(wrapper.find('[data-test="total-price"]').text()).toBe('31.48€');
  });

  it('maches', () => {
    const wrapper = mount(CartView);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
