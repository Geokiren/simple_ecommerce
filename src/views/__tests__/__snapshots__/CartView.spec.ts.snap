// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartView > maches 1`] = `
"<section class="max-w-screen-2xl mx-auto px-6 py-6 flex flex-col gap-4" data-test="cart-view">
  <h1 class="text-2xl text-center font-bold px-4 pt-4 pb-2 w-full border-b">Your Bag</h1>
  <div class="text-sm text-center">Your bag is empty</div>
  <div class="p-4 flex items-center justify-between">
    <div class="font-bold">Subtotal:</div>
    <div data-test="total-price">0â‚¬</div>
  </div>
</section>"
`;
